### ISSUE TITLE
Guidance for changing email when requiring email verification

### ISSUE THREAD

**Issue by madmunsterdaddy** [aspnet/Identity#727]
I would like to see some documentation or template guidance on dealing with a user changing their email address when requiring email's be confirmed :)

**Comment by brockallen** [comment]
All IMO. 

Email verification is not trivial.

Before an account should be allowed to login you should confirm their email address. This motivates the user to do this, because if you don't require it, then they'll never do it. Email being confirmed and trusted is vital for when the password needs to be reset.

To verify an email, you send a code to the email. When this code is then presented to be verified, then the user should be required to re-enter their password to really confirm that the email address the code was sent to is really associated with the user. Email typos happen all the time, and without this the email owner is in a position to pwn the account owner. The prior versions of ASP.NET Identity did not require this additional password step. I feel this is fundamental to building a secure application. If you agree, then you need to add this code/logic on top of ASP.NET Identity. I've not yet spent the time in ASP.NET Identity 3 to know if now has this support.

You must first confirm an email (as above) before it can be changed (duh). To change an email, a code must be sent to the new email, and the same process must be performed as above (meaning the user must re-enter the password to confirm it was the right email). The problem here is that you don't want to overwrite the old email until the new email is confirmed. ASP.NET Identity previously did not have a built-in way to do this, so you needed to build your own infrastructure for this. I've not yet spent the time in ASP.NET Identity 3 to see if they have an easier approach to this.

So that's some guidance from the community. As I mentioned, since many of these important security aspects aren't required or built-in to ASP.NET Identity it will be interesting to see what the proposals are for how to handle them in ASP.NET Identity 3.

**Comment by madmunsterdaddy** [comment]
Yes, I tried to implement this today and failed which prompted this ticket. I created a PendingEmail column on my user class. I generated the code and shot the email to the new email address. I then logged the user out. So far so good. Once they clicked the link in the email I then set the UserName and Email (which in my case are the same), but upon attempting to confirm it, the user manager threw an exception because it didn't match. I guess I could just update that field manually? I don't know. In any case, it seemed like it would be a good scenario to cover ;)
